[10.0][IMP] shopinvader: Improve add item service

rousseldenis opened the pull request:
If there are many things to compute on a sale order,
the item addition to cart could become quite long.

This prevent recomputing cart at each addition.

Adds an option on the backend to enable that behaviour.

To let the cart in a consistent state, recompute
at each time we call get() if needed.

In 'simple' mode, the add_item() service will return :

```python
{"data": {"id": <cart_id>, "lines": {"count" : <cart qty>}}}
```

We keep the format used in normal behaviour to limit the impact on frontend modifications.

Modifications in :
* [x] shopinvader
* [x] shopinvader_promotion_rule

rousseldenis committed:
[10.0][IMP] shopinvader: Improve add item service

If there are many things to compute on a sale order,
the item addition to cart could become quite long.

This prevent recomputing cart at each addition.

To let the cart in a consistent state, recompute
at each time we call get() if needed.

simahawk commented on shopinvader/models/sale.py:
```diff
@@ -45,6 +45,7 @@ class SaleOrder(models.Model):
         compute="_compute_shopinvader_state",
         store=True,
     )
+    shopinvader_to_be_recomputed = fields.Boolean()
```
do we really need a field?
It seems that is used only in the cart service: can't we rely on a ctx key?

rousseldenis replied:
This is a try. I explore possibilities to get quick response from add_item service.

IMHO using context when you try to improve performance is a bad idea.

simahawk replied:
> IMHO using context when you try to improve performance is a bad idea.

why? you need a flag to control a behavior. There's nothing really special.
It doesn't really matter where the flag comes from as long its state is reliable.
If you say "there are other cases where this won't work like 1,2,3" that's another story.

rousseldenis replied:
It's better to use parameters as changing context is creating new environment (< v13)

rousseldenis commented:
First reworked proposal.

I removed the cart recomputation as accessing cart record will recompute it automatically by the orm.

The difficulty is to test values after add_item as ... they are recomputed at field call.

rousseldenis committed:
[10.0][IMP] shopinvader_promotion_rule: Apply on _get

To improve performances on add_item service, we do promotion apply
only on _get

rousseldenis committed:
[10.0][IMP] shopinvader: Delay cart recomputation for add item

rousseldenis committed:
[10.0][IMP] shopinvader_promotion_rule: Delay apply promotion

rousseldenis committed:
[10.0][IMP] shopinvader: Keep info if cart should be recomputed

Add information on cart if it should be recomputed.

rousseldenis commented:
Delayed cart recomputation (in a job) and placed that information in sale.order as field. If the job is failing, depending on the value, the sale order form will display a button to force recomputation.

rousseldenis committed:
[10.0][IMP] shopinvader_delivery_carrier: Test change

As add_item service returns limited information,
do a search to retrieve complete cart.

Cedric-Pigeon commented on shopinvader/services/cart.py:
```diff
@@ -197,16 +225,21 @@ def _clear_cart(self, cart):
+        # Recompute cart asynchronously to avoid latencies on frontend
+        description = "Recompute cart %s" % (existing_item.id)
+        existing_item.order_id.with_delay(
+            description=description
```
I would set priority=1 by default

rousseldenis replied:
Done

rousseldenis committed:
[10.0][IMP] shopinvader: Don't break original flow

rousseldenis committed:
[10.0][IMP] shopinvader: Set delayed add item priority

rousseldenis committed:
[10.0][IMP] shopinvader: Use identity_key for created jobs

As job to recompute could haven't been run between two calls
of 'add_item', don't create new job in that case.

rousseldenis committed:
[10.0][IMP] shopinvader: Protect the whole add_item process

Ensure that during the whole add_item process, the norecompute is applied
not to trigger any recomputation.

For simple add_item service, collect all computed and stored fields,
call _recompute_todo() and then recompute().

rousseldenis committed:
[10.0][IMP] shopinvader,shopinvader_promotion_rule: Recompute on get

Just after an add_item call in simple service mode and if a cart has to be
displayed before the recomputation job has been done, call the recompute().

Add a method to be inherited that allow to not test the field
'shopinvader_to_be_recomputed' at each time.

rousseldenis committed:
[10.0][IMP] shopinvader: Use shopinvader_response

Better use of shopinvader_response in tests

Same behaviour for responses in case of simple add_item

Cedric-Pigeon commented:
@simahawk @lmignon can we go ahead and merge this one. It is used in production since 3 weeks now without any troubles

lmignon commented on shopinvader/models/sale.py:
```diff
@@ -45,6 +46,42 @@ class SaleOrder(models.Model):
     )
+    shopinvader_to_be_recomputed = fields.Boolean(
+        help="Technical field that will intend to check if the sale order has"
+        "to be recomputed du to a modification (asynchronous)"
```
s/du/due

Cedric-Pigeon replied:
done

lmignon commented on shopinvader/services/cart.py:
```diff
@@ -44,10 +50,23 @@ def update(self, **params):
+        self._launch_cart_recompute(cart, item)
+        if simple_service:
+            return self._to_json(cart, simple=True)
         return self._to_json(cart)
```
```suggestion
        return self._to_json(cart, simple=simple_service)
```

Cedric-Pigeon replied:
done

lmignon reviewed (approved):
LGTM (code review only)

simahawk commented on shopinvader/models/shopinvader_backend.py:
```diff
@@ -100,6 +100,10 @@ class ShopinvaderBackend(models.Model):
     )
+    simple_cart_service = fields.Boolean(
+        help="Technical field to change cart service behaviour. Change this"
+        "only if you know what you are doing"
```
I would explain so that ppl know what they can do and which effects this flag has ;)

Cedric-Pigeon replied:
done

Cedric-Pigeon commented:
@simahawk Just improved following your suggestions

Cedric-Pigeon commented:
/ocabot merge minor

sebastienbeau commented on shopinvader/services/cart.py:
```diff
@@ -351,20 +386,39 @@ def _get_step_from_code(self, code):
+            response.set_store_cache("cart", {})
+            return {"data": {}}
+        if simple:
+            res = self._to_json_simple(cart)
```
I think we should return a flag to say explicitely if the cart is simple or not.
simple: true/false

Indeed, on frontend side when you do a "store.cart" you may have a simple or full version (depending if the last call was an update/add or a get)
@rousseldenis

rousseldenis replied:
Ok. Let's add this. Maybe after the merge on another PR ?

Cedric-Pigeon replied:
yes Denis in another PR please. it makes sense to be more generic

sebastienbeau replied:
yes other PR will be perfect
